### PR TITLE
46 auth midware

### PR DESCRIPTION
**Pull request by peacebonez**

### CODE REVIEW

**Comment by ThinkTink** [comment]
Looks like there's a lot of overlap in the files between this PR and the other one you have open. Can you clean it up so each only show the files for that issue, and fix the merge conflict? Thanks!

**Comment by peacebonez** [comment]
Would it ever be a reasonable thing to just delete all files in the branch except for the one being focused on to make merging easier?

**Comment by peacebonez** [comment]
took care of conflicts

**Comment by ThinkTink** [comment]
> Would it ever be a reasonable thing to just delete all files in the branch except for the one being focused on to make merging easier?

Unfortunately, that will just have your branch delete those files on dev when it merges. Is the problem that you're branching off of something other than dev, or that you are using something like `git add .` instead of just selecting the specific relevant files to commit?

File: client/src/contexts/userContext.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  const actions = {
+    fetchUser: async () => {
+      try {
+        const res = await axios.get(`/user/get_by_id/${testUserId}`);
+
+        if (res.status === 200) {
+          const data = await res.data;
+          console.log('res:', res);
+          dispatch({ type: UPDATE_USER, payload: data });
+        } else if (!res || res.status === 400 || res.status === 500) {
```
Comment (ThinkTink) [inline client/src/contexts/userContext.js:29]: We can probably just use a catch-all else here

File: client/src/pages/Landing.js [hunk 2]
```diff
@@ -47,50 +48,109 @@ const useStyles = makeStyles({
 });
 
 export default function Landing() {
+  let history = useHistory();
```
Comment (ThinkTink) [inline client/src/pages/Landing.js:51]: const

File: client/src/pages/Landing.js [hunk 3]
```diff
@@ -47,50 +48,109 @@ const useStyles = makeStyles({
...
-  const [language, setLanguage] = useState('');
+  const [primaryLanguage, setPrimaryLanguage] = useState('');
 
-  const [errorLanguage, setErrorLanguage] = useState('');
+  const [errorName, setErrorName] = useState(false);
+  const [errorEmail, setErrorEmail] = useState(false);
+  const [errorPassword, setErrorPassword] = useState(false);
+  const [errorLanguage, setErrorLanguage] = useState(false);
+  const [backendError, setBackendError] = useState(false);
+  const [backendErrorMsg, setBackendErrorMsg] = useState('');
```
Comment (ThinkTink) [inline client/src/pages/Landing.js:66]: Are you using this?

File: client/src/pages/Landing.js [hunk 4]
```diff
@@ -47,50 +48,109 @@ const useStyles = makeStyles({
...
+      }, 1000);
+    }
+
+    if (!primaryLanguage) {
+      setErrorLanguage(true);
+      let timer = setTimeout(() => {
+        setErrorLanguage(false);
+      }, 1000);
+    }
+  };
```
Comment (ThinkTink) [inline client/src/pages/Landing.js:109]: Is there a way we can handle the errors with less code and without using setTimeout?

File: client/src/pages/Landing.js [hunk 5]
```diff
@@ -47,50 +48,109 @@ const useStyles = makeStyles({
...
-        .post(`${baseURL}/user/signup`, {
-          name, // name: name (shorthand)
+  const handleSubmit = async () => {
+    handleErrorTimouts();
+
+    if (
+      isEmail(email) &&
+      password.length >= 6 &&
+      primaryLanguage &&
+      isName(name)
```
Comment (ThinkTink) [inline client/src/pages/Landing.js:118]: Let's have this check built into the other error checking we're already doing

File: client/src/pages/Landing.js [hunk 6]
```diff
@@ -47,50 +48,109 @@ const useStyles = makeStyles({
...
+    ) {
+      let res = await signUpUser();
```
Comment (ThinkTink) [inline client/src/pages/Landing.js:120]: const

File: client/src/pages/Landing.js [hunk 7]
```diff
@@ -47,50 +48,109 @@ const useStyles = makeStyles({
...
+      if (res && (res.status === 200 || res.status === 201)) {
+        history.push('/messenger');
+      }
+    } else {
+      //TODOS add error snackbar
```
Comment (ThinkTink) [inline client/src/pages/Landing.js:125]:
```suggestion
```

File: client/src/pages/Landing.js [hunk 8]
```diff
@@ -47,50 +48,109 @@ const useStyles = makeStyles({
...
+    }
+  };
+
+  //POST config header values
+  const config = {
+    headers: {
+      'Content-Type': 'application/json',
+    },
+  };
+  const signUpUser = async () => {
```
Comment (ThinkTink) [inline client/src/pages/Landing.js:136]: Since we have userContext now, we should put this function there

File: client/src/pages/Landing.js [hunk 9]
```diff
@@ -47,50 +48,109 @@ const useStyles = makeStyles({
...
+    try {
+      let res = await axios.post(
```
Comment (ThinkTink) [inline client/src/pages/Landing.js:138]: const

File: client/src/pages/Login.js [hunk 10]
```diff
@@ -44,61 +39,67 @@ export default function Login() {
...
-      password
-    }).then(response => {
-      history.push('/messenger');
-    }).catch(error => console.error(error))
+    axios
+      .post(`/user/login`, {
+        email,
+        password,
+      })
+      .then((response) => {
```
Comment (ThinkTink) [inline client/src/pages/Login.js:47]: We want to check the success of the response first, and also move the axios call into userContext

File: models/User.js [hunk 11]
```diff
@@ -5,38 +5,48 @@ const Schema = mongoose.Schema;
...
+        delete ret.password;
+      },
+    },
+    toObject: {
+      transform: function (doc, ret) {
+        ret.id = ret._id;
+        delete ret._id;
+        delete ret.password;
+      },
+    },
```
Comment (ThinkTink) [inline models/User.js:47]: Do we need both of these? Is there a difference between them?

File: routes/invitation.js [hunk 12]
```diff
@@ -2,9 +2,10 @@ const express = require('express');
 const router = express.Router();
 const Invitation = require('../models/Invitation');
 const User = require('../models/User');
+const auth = require('../middleware/auth');
 
-//PUT request to approve invitation
-router.put('/:id/approve', async (req, res) => {
+//PUT request to approve invitation PRIVATE ROUTE
+router.put('/:id/approve', auth, async (req, res) => {
```
Comment (ThinkTink) [inline routes/invitation.js:8]: We also need to add a check that the user who is authenticated is the user whose invitation will be approved, and similar for all protected routes

File: routes/invitation.js [hunk 13]
```diff
@@ -57,6 +59,15 @@ router.put('/:id/approve', async (req, res) => {
         dateJoined: sender.dateJoined,
       };
 
+      // exit if received is already in user's contacts
+      if (
+        sender.contacts.find(
+          (contact) => contact.email.toString() === receiver.email,
+        )
+      ) {
+        return res.status(400).json({ msg: 'Already in contacts' });
```
Comment (ThinkTink) [inline routes/invitation.js:68]: Should this check be done when the invitation is created rather than when it's approved?

File: routes/user.js [hunk 14]
```diff
@@ -53,64 +53,111 @@ router.post(
...
 
-    // Email must be unique
-    const user1 = await User.findOne({ email: req.body.email });
-    if (user1) {
-      return res.status(400).send('User with this email already exists.');
+    const { name, email, password, primaryLanguage } = req.body;
+
+    try {
+      // Email must be unique
+      let user = await User.findOne({ email });
```
Comment (ThinkTink) [inline routes/user.js:70]: const

File: routes/user.js [hunk 15]
```diff
@@ -53,64 +53,111 @@ router.post(
...
+      jwt.sign(
+        payload,
+        process.env.ACCESS_TOKEN_SECRET,
+        { expiresIn: JWT_EXPIRY_TIME },
+        (err, token) => {
+          if (err) throw err;
+          return (
+            res
+              .status(201)
+              // .cookie(token, { httpOnly: true })
```
Comment (ThinkTink) [inline routes/user.js:108]: Shouldn't we keep httpOnly?

File: routes/user.js [hunk 16]
```diff
@@ -53,64 +53,111 @@ router.post(
...
 router.get(
   '/get_by_id/:id',
   runAsyncWrapper(async function (req, res) {
     const user = await User.findById(req.params.id);
-    return res.status(200).json(user);
+
+    if (!user) {
+      return res.status(400).json({ error: 'User not found' });
+    } else {
+      return res.status(200).json(user);
```
Comment (ThinkTink) [inline routes/user.js:131]: We probably don't want to include the user's password in the return here

File: routes/user.js [hunk 17]
```diff
@@ -53,64 +53,111 @@ router.post(
...
+    }
   }),
 );
 
 router.get(
   '/get_all',
   runAsyncWrapper(async function (req, res) {
     const users = await User.find();
-    return res.status(200).json(users);
+    return res.status(200).jsonp(users);
```
Comment (ThinkTink) [inline routes/user.js:155]: typo?

**Review by ThinkTink — COMMENTED** [review]
fyi, some of my comments apply in multiple places, and I didn't write them out in every spot, like needing to verify that it's the right user for each protected route

File: routes/user.js (hunk 16 above)
Comment (peacebonez) [reply routes/user.js:131]: This is addressed at the root in the Users Model.   Bonnie added a .toJson and .toObject transformation method that deletes the password when a user is being retrieved

Comment (ThinkTink) [reply routes/user.js:131]: Oh, I didn't realize that was called automatically with .json(user). :+1:

**Pull request closed without merging**